lib/route/mdb.c

```diff
@@ -0,0 +1,289 @@
+/*
+ * lib/route/mdb.c		Multicast Database
+ *
+ *	This library is free software; you can redistribute it and/or
+ *	modify it under the terms of the GNU Lesser General Public
+ *	License as published by the Free Software Foundation version 2.1
+ *	of the License.
+ *
+ * Copyright (c) 2003-2012 Thomas Graf <tgraf@suug.ch>
+ * Copyright (c) 2003-2006 Baruch Even <baruch@ev-en.org>,
+ *                         Mediatrix Telecom, inc. <ericb@mediatrix.com>
+ */
+
+#include <netlink-private/netlink.h>
+#include <netlink/netlink.h>
+#include <netlink/route/mdb.h>
+#include <netlink/utils.h>
+
+static struct nl_cache_ops rtnl_mdb_ops;
+static struct nl_object_ops mdb_obj_ops;
+/** @endcond */
+
+static void mdb_constructor(struct nl_object *obj)
+{
+  struct rtnl_mdb *_mdb = (struct rtnl_mdb *)obj;
+
+  nl_init_list_head(&_mdb->mdb_entry_list);
+}
+
+static struct nla_policy mdb_policy[MDBA_MAX+1] = {
+	[MDBA_MDB]	= { .type = NLA_NESTED },
+};
+
+static struct nla_policy mdb_db_policy[MDBA_MDB_MAX+1] = {
+	[MDBA_MDB_ENTRY]	= { .type = NLA_NESTED },
+};
+
+static struct nla_policy mdb_entry_policy[MDBA_MDB_ENTRY_MAX+1] = {
+	[MDBA_MDB_ENTRY_INFO]	= { .type = NLA_UNSPEC },
+};
+
+static int mdb_msg_parser(struct nl_cache_ops *ops, struct sockaddr_nl *who,
+			   struct nlmsghdr *nlh, struct nl_parser_param *pp)
+{
+  int err = 0;
+  struct nlattr *tb[MDBA_MAX+1];
+  struct br_port_msg *_port;
+	struct br_mdb_entry* entry;
+
+  struct rtnl_mdb* _mdb = rtnl_mdb_alloc();
+	if (!_mdb)
+		return -NLE_NOMEM;
+
+	err = nlmsg_parse(nlh, sizeof(struct br_port_msg), tb, MDBA_MAX, mdb_policy);
+	if (err < 0)
+		goto errout;
+
+  _mdb->ce_msgtype = nlh->nlmsg_type;
+  _port = nlmsg_data(nlh);
+
+  _mdb->ifindex = _port->ifindex;
+  _mdb->family = _port->family;
+
+  if(tb[MDBA_MDB]) {
+    struct nlattr *db_attr[MDBA_MDB_MAX+1];
+
+    nla_parse_nested(db_attr, MDBA_MDB_MAX, tb[MDBA_MDB], mdb_db_policy);
+
+    if(db_attr[MDBA_MDB_ENTRY]) {
+      struct nlattr *entry_attr[MDBA_MDB_ENTRY_MAX+1];
+      struct rtnl_mdb_entry *_nentry = rtnl_mdb_entry_alloc();
+      
+      nla_parse_nested(entry_attr, MDBA_MDB_ENTRY_MAX, db_attr[MDBA_MDB_ENTRY], mdb_entry_policy);
+
+      entry = nla_data(entry_attr[MDBA_MDB_ENTRY_INFO]);
+
+      _nentry->ifindex = entry->ifindex;
+      _nentry->vid = entry->vid;
+      _nentry->state = entry->state;
+
+      _nentry->proto = ntohs(entry->addr.proto);
+
+      if (_nentry->proto == ETH_P_IP) {
+          char _mcast_address[INET_ADDRSTRLEN]= "";
```

Comment:
Let's not use leading `_` for local variables. That is not following the current style (and does not seem a common practice to me).